src/lib/storage/disk_analyzer.rb

```diff
@@ -0,0 +1,368 @@
+      # (the medium we just booted from to start the installation).
+      #
+      # This is limited with DiskAnalyzer::disk_check_limit because some
+      # architectures (s/390) tend to have a large number of disks, and
+      # checking if a disk is an installation disk involves mounting and
+      # unmounting each partition on that disk.
+      #
+      # @return [Array<::Storage::Disk>] installation disks
+      #
+      def find_installation_disks
+        disks = StorageManager.instance.probed.all_disks.to_a
+        # FIXME: to_a should not be necessary:
+        # libstorage should return something that Ruby can handle.
+        # This is very likely a problem of the Swig bindings.
+
+        usb_disks, non_usb_disks = disks.partition { |disk| disk.transport == ::Storage::USB }
+        disks = usb_disks + non_usb_disks
+        disk_count = 0
+        installation_disks = []
+
+        disks.each do |disk|
+          installation_disks << disk if installation_disk?(disk)
+          disk_count += 1
+          if disk_count > @disk_check_limit
+            log.info("Disk check limit reached after #{disk}")
+            break
+          end
+        end
+        installation_disks
+      end
+
+      # Find disks that are suitable for installing Linux.
+      # Put any USB disks to the end of that array.
+      #
+      # @return [Array<::Storage::Disk>] candidate disks
+      #
+      def find_candidate_disks
+        disks = StorageManager.instance.probed.all_disks.to_a
+        # FIXME: to_a should not be necessary
+
+        usb_disks, non_usb_disks = disks.partition { |disk| disk.transport == ::Storage::USB }
+        log.info("USB Disks:     #{dev_names(usb_disks)}")
+        log.info("Non-USB Disks: #{dev_names(non_usb_disks)}")
+
+        # We don't want to install on our installation disk if there is any other way.
+        candidate_disks = remove_installation_disks(non_usb_disks + usb_disks)
+        if candidate_disks.empty?
+          log.info("No candidate disks left after eliminating installation disks")
+          log.info("Trying with non-USB installation disks")
+          candidate_disks = @installation_disks.select { |disk| disk.transport != ::Storage::USB }
+        end
+        if candidate_disks.empty?
+          log.info("Still no candidate disks left")
+          log.info("Trying with installation disks out of sheer desperation")
+          candidate_disks = @installation_disks
+        end
+        candidate_disks
+      end
+
+      # Remove any installation disks from 'disks' and return a disks array
+      # containing the disks that are not installation media.
+      #
+      # @param disks [Array<::Storage::Disk>]
+      # @return [Array<::Storage::Disk>] non-installation disks
+      #
+      def remove_installation_disks(disks)
+        # We can't simply use
+        #   disks -= @installation_disks
+        # because the list elements (from libstorage) don't provide a .hash method.
```

Comment:
Yes, if you know how to do that?

IIRC Arvin successfully expanded that ::Storage class with a Ruby extension. Not sure exactly what is needed for that, though.